src/Security/Authentication/Certificate/src/CertificateValidationCache.cs

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Security.Cryptography.X509Certificates;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Caching.Memory;
+
+namespace Microsoft.AspNetCore.Authentication.Certificate
+{
+    /// <summary>
+    /// MemoryCache based implementation used to store <see cref="AuthenticateResult"/> results after the certificate has been validated
+    /// </summary>
+    public class CertificateValidationCache : ICertificateValidationCache
+    {
+        private MemoryCache _cache = new MemoryCache(new MemoryCacheOptions { SizeLimit = 1024 });
+
+        /// <summary>
+        /// Get the <see cref="AuthenticateResult"/> for the connection and certificate.
+        /// </summary>
+        /// <param name="connection">The connection.</param>
+        /// <param name="certificate">The certificate.</param>
+        /// <returns>the <see cref="AuthenticateResult"/></returns>
+        public AuthenticateResult Get(ConnectionInfo connection, X509Certificate certificate)
+            => _cache.Get(ComputeKey(connection, certificate)) as AuthenticateResult;
+
+        /// <summary>
+        /// Store a <see cref="AuthenticateResult"/> for the connection and certificate
+        /// </summary>
+        /// <param name="connection">The connection.</param>
+        /// <param name="certificate">The certificate.</param>
+        /// <param name="result">the <see cref="AuthenticateResult"/></param>
+        public void Put(ConnectionInfo connection, X509Certificate certificate, AuthenticateResult result)
```

Comment:
The auth result shouldn't be shared across requests, no? The Principal is mutable and something like ClaimsTransformation could mess it up. Would you have to do a deep clone when putting it into the cache?

Reply:
Yeah I guess we should use the IPrincipal ctor

Reply:
Added a clone to AuthenticateResult/Ticket to do the deep copy

Reply:
Added a clone to AuthenticateResult/Ticket to do the deep copy